clicommand/pipeline_upload.go

```diff
@@ -283,25 +304,68 @@ var PipelineUploadCommand = cli.Command{
 			searchForSecrets(l, &cfg, envMap, result, src)
 		}
 
-		if cfg.SigningKeyPath != "" {
+		if cfg.JWKSFilePath != "" {
 			l.Warn("Pipeline signing is experimental and the user interface might change! Also it might not work, it might sign the pipeline only partially, or it might eat your pet dog. You have been warned!")
 
-			key, err := os.ReadFile(cfg.SigningKeyPath)
+			jwksFile, err := os.Open(cfg.JWKSFilePath)
 			if err != nil {
-				l.Fatal("Couldn't read the signing key file: %v", err)
+				l.Fatal("Couldn't open JWKS file: %v", err)
 			}
+			defer jwksFile.Close()
 
-			// TODO: Let the user choose an algorithm, then parse the key based
-			// on the algorithm, or put key parsing into
-			// pipeline.New{Signer,Verifier}.
-			// For now we only offer hmac-sha256, which takes []byte.
+			jwksBody, err := io.ReadAll(jwksFile)
+			if err != nil {
+				l.Fatal("Couldn't read JWKS file: %v", err)
+			}
 
-			signer, err := pipeline.NewSigner("hmac-sha256", key)
+			jwks, err := jwk.Parse(jwksBody)
 			if err != nil {
-				l.Fatal("Couldn't create a pipeline signer: %v", err)
+				l.Fatal("Couldn't parse JWKS file: %v", err)
 			}
 
-			if err := result.Sign(signer); err != nil {
+			if cfg.SigningKeyID == "" {
+				l.Fatal("Signing key ID is required when using JWKS")
+			}
+
+			key, found := jwks.LookupKeyID(cfg.SigningKeyID)
+			if !found {
+				l.Fatal("Couldn't find signing key ID %q in JWKS", cfg.SigningKeyID)
+			}
+
+			var alg jwa.KeyAlgorithm
+
+			if key.Algorithm().String() == "" {
+				if cfg.SigningAlgorithm == "" {
+					l.Fatal("Signing algorithm is required if JWKS key doesn't have an alg parameter")
+				}
+				alg = jwa.KeyAlgorithmFrom(cfg.SigningAlgorithm)
+			} else {
+				alg = key.Algorithm()
+			}
+
+			if _, ok := alg.(jwa.InvalidKeyAlgorithm); ok {
+				l.Fatal("Signing algorithm %s is not a valid signing algorithm", alg)
+			}
+
+			signingAlg, ok := alg.(jwa.SignatureAlgorithm)
+			if !ok {
+				var from string
+				if key.Algorithm().String() != "" {
+					from = fmt.Sprintf("from JWKS, key ID %s", cfg.SigningKeyID)
+				} else {
+					from = "from --signing-algorithm flag"
+				}
+
+				l.Fatal("Algorithm %s (%s) cannot be used for signing/verification", signingAlg, from)
+			}
+
+			if err := validateJWK(key, signingAlg); err != nil {
+				l.Fatal("Signing key ID %s is invalid: %v", cfg.SigningKeyID, err)
+			}
+
+			key.Set(jwk.AlgorithmKey, signingAlg)
```

Comment:
this algorithm wrangling might be the tricksiest part of the PR - currently, we allow users to either specify an algorithm in their JWK (under the `alg` key, supported by JWK), or allow them to specify an algorithm by passing the `--signing-algorithm` cli arg.

whatever happens, we end up validating that the key they've given us is capable of using that algorithm, then, if it wasn't already set, we set the key's `alg` to what they've passed us. it's a little complicated, but when we sign the pipeline later on, it allows us to just call `jws.Sign([]byte{"blah"}, jws.WithKey(key.Algorithm(), key)` rather than sidechanneling it all the way through. happy to update this logic, as it's a little bit complex